### ISSUE TITLE
[NEXT-1092] `next build` compile error regarding export worker

### ISSUE THREAD

**Issue by Kazzuk** [vercel/next.js#49059]
### Verify canary release

* [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 11 Pro
Binaries:
  Node: 18.16.0
  npm: 9.5.1
  Yarn: N/A
  pnpm: 8.4.0
Relevant packages:
  next: 13.4.1-canary.2
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true), Package manager (npm, pnpm, Yarn), Static HTML Export (output: "export")

### Link to the code that reproduces this issue

[https://github.com/Kazzuk/reproduce](https://github.com/Kazzuk/reproduce)

### To Reproduce

* Ensure `output` is set to "export" in `next.config.js` to cause error.
* Run `pnpm build`.

### Describe the Bug

Whenever `output` is set to "export" in `next.config.js` and you build the project you will be met with the below error. If I was to remove `output` from `next.config.js` the project compiles fine. As well if remove the dynamic routes (delete `blog` folder) from the project while `output` is set to "export" the project compiles fine.

```
> Build error occurred
Error: invariant: Undefined export worker for app dir, this is a bug in Next.js.
    at C:\Users\Kazuk\Documents\GitHub\reproduce\node_modules\.pnpm\next@13.4.1-canary.2_react-dom@18.2.0_react@18.2.0\node_modules\next\dist\export\index.js:506:27
    at Span.traceAsyncFn (C:\Users\Kazuk\Documents\GitHub\reproduce\node_modules\.pnpm\next@13.4.1-canary.2_react-dom@18.2.0_react@18.2.0\node_modules\next\dist\trace\trace.js:103:26)
    at C:\Users\Kazuk\Documents\GitHub\reproduce\node_modules\.pnpm\next@13.4.1-canary.2_react-dom@18.2.0_react@18.2.0\node_modules\next\dist\export\index.js:502:35
    at Array.map (<anonymous>)
    at C:\Users\Kazuk\Documents\GitHub\reproduce\node_modules\.pnpm\next@13.4.1-canary.2_react-dom@18.2.0_react@18.2.0\node_modules\next\dist\export\index.js:499:41
    at async Span.traceAsyncFn (C:\Users\Kazuk\Documents\GitHub\reproduce\node_modules\.pnpm\next@13.4.1-canary.2_react-dom@18.2.0_react@18.2.0\node_modules\next\dist\trace\trace.js:103:20)    
    at async C:\Users\Kazuk\Documents\GitHub\reproduce\node_modules\.pnpm\next@13.4.1-canary.2_react-dom@18.2.0_react@18.2.0\node_modules\next\dist\build\index.js:1887:17
    at async Span.traceAsyncFn (C:\Users\Kazuk\Documents\GitHub\reproduce\node_modules\.pnpm\next@13.4.1-canary.2_react-dom@18.2.0_react@18.2.0\node_modules\next\dist\trace\trace.js:103:20)    
    at async build (C:\Users\Kazuk\Documents\GitHub\reproduce\node_modules\.pnpm\next@13.4.1-canary.2_react-dom@18.2.0_react@18.2.0\node_modules\next\dist\build\index.js:143:29)
```

### Expected Behavior

Build should be successful.

### Which browser are you using? (if relevant)

*No response*

### How are you deploying your application? (if relevant)

*No response*

<sub>From [SyncLinear.com](https://synclinear.com) | [NEXT-1092](https://linear.app/vercel/issue/NEXT-1092/next-build-compile-error-regarding-export-worker)</sub>

**Comment by Kazzuk** [comment]
@shuding This commit b5f7f8448560d4ec6bc875cbffe3aa18da8971cf may have something to do with this.

**Comment by mtnowl** [comment]
It also seems to go away when you put `getStaticParams` in dynamic route pages.

**Comment by alarcndaniela** [comment]
I'm having the same issue

**Comment by detachedsoul** [comment]
I'm also having the same issue.

**Comment by rjruizes** [comment]
In my case, fixing my `generateStaticParams` functions fixed the error. I have 3 dynamic segments in my URL, e.g. `/[First]/[Second]/[Third]`. 
```
[org]/layout.tsx: generateStaticParams returns array for possible values of `org`
  [dept]/layout.tsx: generateStaticParams returns array for possible values of `dept`
    [unit]/page.tsx: generateStaticParams returns array for possible values of `unit`
```

Doc: https://beta.nextjs.org/docs/api-reference/generate-static-params

Initially I had incorrect values in `[unit]/page.tsx` and I isolated the problem by removing files one at a time.

**Comment by mtnowl** [comment]
@rjruizes Those child `generateStaticParams` calls are working for you? I'm getting [this bug](https://github.com/vercel/next.js/issues/49078) where the child `generateStaticParams` are being passed `undefined` parameters.

**Comment by alarcndaniela** [comment]
@rjruizes This is definitely the root cause for the error in my case, yesterday I moved the files one at a time to a new project and discovered it. Thank you!

**Comment by rjruizes** [comment]
@mtnowl The params are defined for me. I pushed up an example repo to demo it working, see [page.tsx](https://github.com/rjruizes/nextjs-with-tailwindcss-app/blob/main/src/app/%5Borg%5D/%5Bdept%5D/%5Bunit%5D/page.tsx).

**Comment by alarcndaniela** [comment]
I just discovered that in my case everytime I add a folder estructure like /dashboard/admin/companies/[id] the error pops up

**Comment by ytsruh** [comment]
Same issue from me. Unable to use build / export whilst using a dyanmic route and client side component.

Error: invariant: Undefined export worker for app dir, this is a bug in Next.js.

**Comment by alarcndaniela** [comment]
I stoped using the last version of nextjs 13 with the experimental app directory and changed it to a project without app and src directory and the build was successfully made

**Comment by mdmathewdc** [comment]
I am also facing the same issue with Next 13.4.1. The document suggests`generateStaticParams` method to statically generate a dynamic route (https://nextjs.org/docs/app/api-reference/functions/generate-static-params).

But is there an option to make it dynamic if I don't know the `props` at build time? eg: In my application, I'm generating a unique UUID so that it can be shared as a link - `https://xyz.app/play/UUID`. How will export the app as a static site then?

**Comment by brandensilva** [comment]
The same issue is occurring for me on 13.4.1 but I am doing nothing with params. Only one server side fetch going on and its happening.

**Comment by Kazzuk** [comment]
Bumped the reproduction [repository](https://github.com/Kazzuk/reproduce) to latest canary (`13.4.1-canary.2`) as of the moment, also changed issue description to reflect error log.

**Comment by mkarajohn** [comment]
I have this issue as well.

## Context

I am moving an existing CRA project over to Next. I have set up the `app` router so that I have a single catch-all root, like this:

```
src/
 |-app/
   |-[[...path]]/
   | |-page.tsx
   |-layout.tsx
```

I am trying to build it as a SPA using the `output: 'export'` option in the config.

The build fails and the `out` directory ends up having only a `404.html` page.


## Error

```
- info Linting and checking validity of types  
- info Collecting page data  
- info Generating static pages (3/3)
- info Finalizing page optimization  

Route (app)                                Size     First Load JS
┌ λ /[[...path]]                           1.4 kB         79.1 kB
└ ○ /favicon.ico                           0 B                0 B
+ First Load JS shared by all              77.7 kB
  ├ chunks/139-f94b65a583fb7f46.js         24.6 kB
  ├ chunks/2443530c-c1df3a32f9287dc4.js    50.5 kB
  ├ chunks/main-app-f3225843db279bbc.js    221 B
  └ chunks/webpack-33cf70d50500edbd.js     2.37 kB

Route (pages)                              Size     First Load JS
─ ○ /404                                   180 B          86.8 kB
+ First Load JS shared by all              86.6 kB
  ├ chunks/main-3a574163a2858406.js        84.1 kB
  ├ chunks/pages/_app-dec1a83570915fb0.js  195 B
  └ chunks/webpack-33cf70d50500edbd.js     2.37 kB

λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
○  (Static)  automatically rendered as static HTML (uses no initial props)


> Build error occurred
Error: invariant: Undefined export worker for app dir, this is a bug in Next.js.
    at /home/mitch/Workspace/next-projects/project/node_modules/next/dist/export/index.js:506:27
    at Span.traceAsyncFn (/home/mitch/Workspace/next-projects/project/node_modules/next/dist/trace/trace.js:103:26)
    at /home/mitch/Workspace/next-projects/project/node_modules/next/dist/export/index.js:502:35
    at Array.map (<anonymous>)
    at /home/mitch/Workspace/next-projects/project/node_modules/next/dist/export/index.js:499:41
    at async Span.traceAsyncFn (/home/mitch/Workspace/next-projects/project/node_modules/next/dist/trace/trace.js:103:20)
    at async /home/mitch/Workspace/next-projects/project/node_modules/next/dist/build/index.js:1887:17
    at async Span.traceAsyncFn (/home/mitch/Workspace/next-projects/project/node_modules/next/dist/trace/trace.js:103:20)
    at async build (/home/mitch/Workspace/next-projects/project/node_modules/next/dist/build/index.js:143:29)
```

## Setup info

```
    Operating System:
      Platform: linux
      Arch: x64
      Version: #41~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Mar 31 16:00:14 UTC 2
    Binaries:
      Node: 18.12.1
      npm: 8.19.2
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.4.1-canary.2
      eslint-config-next: 13.4.1
      react: 18.2.0
      react-dom: 18.2.0
```

**Comment by jayalmaraz** [comment]
+1 getting this error on `13.4.1` & have been getting it since `13.3.2`

```
> Build error occurred
Error: invariant: Undefined export worker for app dir, this is a bug in Next.js.
```

[MCVE](https://github.com/jayalmaraz/20230508-next-13-app-router-static-export-bug)

Obviously `generateStaticParams` isn't a real solution. We are relying on Dynamic Route Segments with Client Component data fetching ([documented supported by Static Export](https://nextjs.org/docs/app/building-your-application/deploying/static-exports#dynamic-route-segments)), and expecting it to work

```
macOS 12.6.2
node 16.20.0
```

**Comment by styfle** [comment]
This is fixed on the canary channel.

You can try it today with `npm install next@canary`, thanks!

**Comment by Yehorchik** [comment]
@styfle i just tried it on `13.4.2-canary.3` and it's still has the same issue :(

**Comment by ytsruh** [comment]
> @styfle i just tried it on `13.4.2-canary.3` and it's still has the same issue :(

Works fine for me

**Comment by styfle** [comment]
@Yehorchik You might have a different issue. See #48022 

@ytsruh Thanks for confirming 👍

**Comment by nhat-em-wick** [comment]
@styfle thanks

**Comment by brandensilva** [comment]
Also confirmed the fix, thanks.

**Comment by oldbettie** [comment]
confirmed fixed in `v13.4.4`
